// umi-ui-tasks/lib/ui/components/Install/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _antd = require("antd");

var _icons = require("@ant-design/icons");

var _reactSizeme = _interopRequireDefault(require("react-sizeme"));

var _uiModule = _interopRequireDefault(require("../../ui.module.less"));

var _enums = require("../../../server/core/enums");

var _util = require("../../util");

var _model = require("../../model");

var _hooks = require("../../hooks");

var _Terminal = _interopRequireDefault(require("../Terminal"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var SizeMe = _reactSizeme.default.SizeMe;
var taskType = _enums.TaskType.INSTALL;
var Option = _antd.Select.Option;

var InstallComponent = function InstallComponent(_ref) {
  var api = _ref.api,
      dispatch = _ref.dispatch,
      _ref$detail = _ref.detail,
      detail = _ref$detail === void 0 ? {} : _ref$detail;
  var intl = api.intl;

  var _useState = (0, _react.useState)(''),
      _useState2 = _slicedToArray(_useState, 2),
      log = _useState2[0],
      setLog = _useState2[1];

  var _useState3 = (0, _react.useState)([]),
      _useState4 = _slicedToArray(_useState3, 2),
      npmClients = _useState4[0],
      setNpmClients = _useState4[1];

  var _useState5 = (0, _react.useState)(false),
      _useState6 = _slicedToArray(_useState5, 2),
      modalVisible = _useState6[0],
      setModalVisible = _useState6[1];

  var _Form$useForm = _antd.Form.useForm(),
      _Form$useForm2 = _slicedToArray(_Form$useForm, 1),
      form = _Form$useForm2[0];

  var _useInit = (0, _hooks.useInit)(detail),
      _useInit2 = _slicedToArray(_useInit, 1),
      init = _useInit2[0];

  (0, _react.useEffect)(function () {
    if (!init) {
      return;
    }

    dispatch({
      type: "".concat(_model.namespace, "/getTaskDetail"),
      payload: {
        taskType: taskType,
        callback: function callback(_ref2) {
          var log = _ref2.log;
          setLog(log);
        }
      }
    });

    _asyncToGenerator(
    /*#__PURE__*/
    regeneratorRuntime.mark(function _callee() {
      var _ref4, data;

      return regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              _context.next = 2;
              return api.callRemote({
                type: '@@project/getNpmClients'
              });

            case 2:
              _ref4 = _context.sent;
              data = _ref4.data;
              setNpmClients(data);
              form.setFieldsValue({
                npmClient: data[0]
              });

            case 6:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }))();
  }, [init]); // UnMount: reset form

  (0, _react.useEffect)(function () {
    return function () {
      form.resetFields();
      var terminal = (0, _util.getTerminalIns)(taskType);
      terminal && terminal.clear();
    };
  }, []);

  function install(_x) {
    return _install.apply(this, arguments);
  }

  function _install() {
    _install = _asyncToGenerator(
    /*#__PURE__*/
    regeneratorRuntime.mark(function _callee2(npmClient) {
      return regeneratorRuntime.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              dispatch({
                type: "".concat(_model.namespace, "/exec"),
                payload: {
                  taskType: taskType,
                  env: {
                    NPM_CLIENT: npmClient
                  }
                }
              });

            case 1:
            case "end":
              return _context2.stop();
          }
        }
      }, _callee2);
    }));
    return _install.apply(this, arguments);
  }

  function cancelInstall() {
    return _cancelInstall.apply(this, arguments);
  }

  function _cancelInstall() {
    _cancelInstall = _asyncToGenerator(
    /*#__PURE__*/
    regeneratorRuntime.mark(function _callee3() {
      return regeneratorRuntime.wrap(function _callee3$(_context3) {
        while (1) {
          switch (_context3.prev = _context3.next) {
            case 0:
              dispatch({
                type: "".concat(_model.namespace, "/cancel"),
                payload: {
                  taskType: taskType
                }
              });

            case 1:
            case "end":
              return _context3.stop();
          }
        }
      }, _callee3);
    }));
    return _cancelInstall.apply(this, arguments);
  }

  var openModal = function openModal() {
    setModalVisible(true);
  };

  var handleOk = function handleOk() {
    form.validateFields().then(function (values) {
      install(values.npmClient);
      setModalVisible(false);
    }).catch(function (_) {});
  };

  var handleCancel = function handleCancel() {
    setModalVisible(false);
  };

  var isTaskRunning = detail && detail.state === _enums.TaskState.ING;
  return _react.default.createElement(_react.default.Fragment, null, _react.default.createElement("h1", {
    className: _uiModule.default.title
  }, intl({
    id: 'org.umi.ui.tasks.install'
  })), _react.default.createElement(_react.default.Fragment, null, _react.default.createElement(_antd.Row, null, _react.default.createElement(_antd.Col, {
    span: 24,
    className: _uiModule.default.buttonGroup
  }, _react.default.createElement(_antd.Button, {
    size: api.mini ? 'small' : 'default',
    type: "primary",
    onClick: isTaskRunning ? cancelInstall : openModal
  }, isTaskRunning ? _react.default.createElement(_react.default.Fragment, null, _react.default.createElement(_icons.Pause, null), _react.default.createElement("span", {
    className: _uiModule.default.runningText
  }, ' ', intl({
    id: 'org.umi.ui.tasks.install.cancel'
  }))) : _react.default.createElement(_react.default.Fragment, null, _react.default.createElement(_icons.CaretRight, null), _react.default.createElement("span", {
    className: _uiModule.default.runningText
  }, ' ', intl({
    id: 'org.umi.ui.tasks.install.start'
  })))), _react.default.createElement(_antd.Modal, {
    visible: modalVisible,
    title: intl({
      id: 'org.umi.ui.tasks.install'
    }),
    onOk: handleOk,
    onCancel: handleCancel
  }, _react.default.createElement("div", {
    className: _uiModule.default.modalContainer
  }, _react.default.createElement("div", {
    className: _uiModule.default.confirmMessage
  }, intl({
    id: 'org.umi.ui.tasks.install.tip'
  })), _react.default.createElement(_antd.Form, {
    name: "intasllEnv",
    form: form,
    layout: "vertical"
  }, _react.default.createElement(_antd.Form.Item, {
    label: intl({
      id: 'org.umi.ui.tasks.install.npmClient'
    }),
    name: "npmClient"
  }, _react.default.createElement(_antd.Select, {
    style: {
      width: 120
    }
  }, npmClients.map(function (key) {
    return _react.default.createElement(Option, {
      key: key,
      value: key
    }, key);
  })))))))), _react.default.createElement("div", {
    className: _uiModule.default.logContainer
  }, _react.default.createElement(SizeMe, {
    monitorWidth: true,
    monitorHeight: true
  }, function (_ref5) {
    var size = _ref5.size;
    return _react.default.createElement(_Terminal.default, {
      api: api,
      size: size,
      terminal: (0, _util.getTerminalIns)(taskType),
      log: log,
      onClear: function onClear() {
        (0, _util.clearLog)(taskType);
      }
    });
  }))));
};

var _default = InstallComponent;
exports.default = _default;
